### PR TITLE
pallet-beefy: ensure mandatory block once per session

### PR DESCRIPTION
**Pull request by acatangiu**
BEEFY mandatory blocks [are defined as](https://github.com/paritytech/grandpa-bridge-gadget/blob/8e7d82917e988d7e54ff2ffecdf8461bb9f65c53/docs/beefy.md?plain=1#L177) one per session, regardless of whether authority set changed.

This PR changes the BEEFY authority set _each_ session, even when the authorities stay the same. A new set with a new `ID` is created each session.

The reasoning for one **mandatory BEEFY block** per session is to make sure we don't go for long periods without any BEEFY justification.

In the particular case of `Lean BEEFY`, this also helps restarted nodes to start voting faster since they will start voting after first witnessed mandatory BEEFY block justification.

Fixes: https://github.com/paritytech/grandpa-bridge-gadget/issues/411

### CODE REVIEW

**Comment by acatangiu** [comment]
## Alternative solution:
Another option would be to keep the BEEFY validator set in sync with GRANDPA, and add new (slimmer) `Beefy::NewSession` or `Beefy::MandatoryBlock` header digest that gets deposited once per session.

[AuthoritiesChange](https://github.com/paritytech/substrate/blob/137e6c3ea4150a81f0dfee3a45189c2d988e6b85/primitives/beefy/src/lib.rs#L134) digest would only be deposited on actual ValidatorSet changes.

### Alternative PROS:
- follows GRANDPA validator set changes, but not seeing any particular advantages in doing so
- somewhat slimmer headers, but inconsequential since we're talking about 1 digest per session

### Alternative CONS:
- losing _Mandatory Blocks_ 1:1 mapping with _Validator Sets_ - in current PR, even if validators within the set doesn't change, the validator set id increases monotonically alongside BEEFY mandatory blocks (and frame sessions) - with the alternative described above, we'd lose that.

### Personal recommendation:
Stay with current PR proposal.

**Review by svyatonik — COMMENTED** [review]
I don't have any questions about the code changes, but can you, please explain here shortly (for those who are out of loop, like me):
- the same document states that the algorithm for BEEFY and GRANDPA is the same: "Since GRANDPA is considering session boundary blocks as mandatory as well, `session_start` block will always have both GRANDPA and BEEFY Justification". But iiuc you're saying that if GRANDPA authorities are not changed, the begin-session header is not mandatory for GRANDPA. Could it be that the document is wrong here and we shall leave everything as-is;
- follow up of ^^^: if the document is correct, then what's the reason to mark begin-session headers as mandatory? What we want to achieve/prove with this justification?;
- dumb q: what is meant by "the underlying economic identities may have changed". Is it validator stake or something else?

**Comment by acatangiu** [comment]
> I don't have any questions about the code changes, but can you, please explain here shortly (for those who are out of loop, like me):

Good point, I've added the info below in the PR description as well.

> the same document states that the algorithm for BEEFY and GRANDPA is the same: "Since GRANDPA is considering session boundary blocks as mandatory as well, `session_start` block will always have both GRANDPA and BEEFY Justification". But iiuc you're saying that if GRANDPA authorities are not changed, the begin-session header is not mandatory for GRANDPA. Could it be that the document is wrong here and we shall leave everything as-is;

I'm not sure if GRANDPA justification is guaranteed for `session_start` block if authorities are not changed in said session. I will find out and update the BEEFY doc if it turns out it's mistaken in this case.

> follow up of ^^^: if the document is correct, then what's the reason to mark begin-session headers as mandatory? What we want to achieve/prove with this justification?;

The reasoning for one **mandatory BEEFY block** per session is to make sure we don't go for long periods without any BEEFY justification.

In the particular case of `Lean BEEFY`, this also helps restarted nodes to start voting faster since they will start voting after first witnessed mandatory BEEFY block justification.

> dumb q: what is meant by "the underlying economic identities may have changed". Is it validator stake or something else?

Right! I think stake changes are relevant too, also I believe you could have new validator elections resulting in the same set as before; even though keys didn't change, a new validator set was in fact elected.

**Comment by andresilva** [comment]
> > 
> 
> I'm not sure if GRANDPA justification is guaranteed for `session_start` block if authorities are not changed in said session. I will find out and update the BEEFY doc if it turns out it's mistaken in this case.

It is not guaranteed, only for blocks that change the authorities (and therefore start a new "set id").

File: frame/beefy-mmr/src/tests.rs [hunk 1]
```diff
@@ -70,9 +70,14 @@ fn should_contain_mmr_digest() {
 
 		assert_eq!(
 			System::digest().logs,
-			vec![beefy_log(ConsensusLog::MmrRoot(
-				hex!("fa0275b19b2565089f7e2377ee73b9050e8d53bce108ef722a3251fd9d371d4b").into()
-			))]
+			vec![
+				beefy_log(ConsensusLog::AuthoritiesChange(
+					ValidatorSet::new(vec![mock_beefy_id(1), mock_beefy_id(2),], 1,).unwrap()
```
Comment (andresilva) [inline frame/beefy-mmr/src/tests.rs:75]:
```suggestion
					ValidatorSet::new(vec![mock_beefy_id(1), mock_beefy_id(2)], 1).unwrap()
```

File: frame/beefy-mmr/src/tests.rs [hunk 2]
```diff
@@ -81,14 +86,17 @@ fn should_contain_mmr_digest() {
...
+				)),
 				beefy_log(ConsensusLog::MmrRoot(
-					hex!("fa0275b19b2565089f7e2377ee73b9050e8d53bce108ef722a3251fd9d371d4b").into()
+					hex!("95803defe6ea9f41e7ec6afa497064f21bfded027d8812efacbdf984e630cbdc").into()
 				)),
 				beefy_log(ConsensusLog::AuthoritiesChange(
-					ValidatorSet::new(vec![mock_beefy_id(3), mock_beefy_id(4),], 1,).unwrap()
+					ValidatorSet::new(vec![mock_beefy_id(3), mock_beefy_id(4),], 2,).unwrap()
```
Comment (andresilva) [inline frame/beefy-mmr/src/tests.rs:96]:
```suggestion
					ValidatorSet::new(vec![mock_beefy_id(3), mock_beefy_id(4)], 2).unwrap()
```

File: frame/beefy/src/lib.rs [hunk 3]
```diff
@@ -105,10 +105,10 @@ impl<T: Config> Pallet<T> {
...
 
 	fn change_authorities(new: Vec<T::BeefyId>, queued: Vec<T::BeefyId>) {
-		// Always issue a change if `session` says that the validators have changed.
-		// Even if their session keys are the same as before, the underlying economic
-		// identities have changed. Furthermore, the digest below is used to signal
-		// BEEFY mandatory blocks.
+		// Always issue a change on each `session`.
+		// Even if the validator session keys are the same as before, the underlying economic
+		// identities may have changed.
+		// Furthermore, the digest below is used to signal BEEFY mandatory blocks.
```
Comment (andresilva) [inline frame/beefy/src/lib.rs:111]: I think I would probably remove this whole comment block now, IMO it adds more confusion. The comment we have below on the `on_new_session` method is probably enough.

**Comment by tomusdrw** [comment]
Note that the spec does not mention the digest items at all. IMHO the description here is a bit misleading, cause it assumes that "mandatory blocks" are the ones with the BEEFY digest. While practically it's probably desirable and that's exactly how it works now, I'd consider this an implementation detail of the gadget code and not the actual spec requirement (BEEFY does not deal with headers at all). I.e. we had mandatory blocks at session boundaries even now, but we fail to "detect" them in the digest.

Might be worth adding some sort of documentation about that to the spec imho.

**Review by svyatonik — APPROVED** [review]

File: frame/beefy/src/lib.rs (hunk 3 above)
Comment (acatangiu) [inline frame/beefy/src/lib.rs:111]:
```suggestion
```

**Comment by acatangiu** [comment]
bot merge

**Pull request merged**